File: java/anagram/src/main/java/Anagram.java
import java.util.List;
import java.util.stream.Collectors;

class Anagram {
    private final String anagram;
    private final String original;

    Anagram(String word) {
        anagram = toAnagram(word);
        original = word;
    }

    private String toAnagram(String word) {
        return word
                .toLowerCase()
                .chars()
                .mapToObj(c -> (char) c)
                .sorted()
                .collect(StringBuilder::new, StringBuilder::append, StringBuilder::append)
                .toString();
    }

    List<String> match(List<String> words) {
        return words
                .stream()
                .filter(w -> !w.equalsIgnoreCase(original))
                .filter(w -> toAnagram(w).equals(anagram))
                .collect(Collectors.toList());
    }
}

File: typescript/anagram/anagram.ts
class Anagram {
  readonly word: string
  private readonly key: string

  constructor(word: string) {
    this.word = word
    this.key = this.toKey(word)
  }

  private toKey(word: string): string {
    return [...word.toLowerCase()].sort().join('')
  }

  matches(...words: string[]): string[] {
    return words
      .filter((word) => word.toLowerCase() !== this.word.toLowerCase())
      .filter((word) => this.toKey(word) === this.key)
  }
}

export { Anagram }
